fix(orm)!: duration encoding doesn't handle nil values properly

aaronc opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Closes: #11980

This changes the encoding of durations in the ORM to the one proposed in #11980.

It also fixes a bug with the timestamp encoding where the min and max seconds values were calculated incorrectly (by me). I found this by adding tests for the min and max cases for both timestamps and durations. It would be good to have a second pair of eyes look closely at both of these encodings - especially the hard coded min/max values to make sure they're correct.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

ryanchristo commented on orm/encoding/ormfield/duration.go:
```diff
@@ -1,11 +1,124 @@
+
+	if negative {
+		if nanosInt < DurationNanosMin || nanosInt > 0 {
+			return fmt.Errorf("negative duration nanos is out of range %d, must be between %d and %d", secondsInt, DurationNanosMin, 0)
```
```suggestion
			return fmt.Errorf("negative duration nanos is out of range %d, must be between %d and %d", nanosInt, DurationNanosMin, 0)
```

ryanchristo replied:
Might be nice to include a test case for expected error.

ryanchristo commented on orm/encoding/ormfield/duration.go:
```diff
@@ -1,11 +1,124 @@
+		nanosInt = -nanosInt
+	} else {
+		if nanosInt < 0 || nanosInt > DurationNanosMax {
+			return fmt.Errorf("duration nanos is out of range %d, must be between %d and %d", secondsInt, 0, DurationNanosMax)
```
```suggestion
			return fmt.Errorf("duration nanos is out of range %d, must be between %d and %d", nanosInt, 0, DurationNanosMax)
```

ryanchristo replied:
Might be nice to include a test case for expected error.

ryanchristo commented on orm/encoding/ormfield/duration.go:
```diff
@@ -1,11 +1,124 @@
+		return fmt.Errorf("duration seconds is out of range %d, must be between %d and %d", secondsInt, DurationSecondsMin, DurationSecondsMax)
+	}
+	negative := secondsInt < 0
+	secondsInt -= DurationSecondsMin
```
A comment might be helpful here explaining why min amount is being subtracted.

ryanchristo commented on orm/encoding/ormfield/duration.go:
```diff
@@ -1,11 +1,124 @@
+		return protoreflect.Value{}, err
+	}
+
+	seconds += DurationSecondsMin
```
Same as above, a comment might be helpful here explaining why min amount is being added.

ryanchristo commented on orm/encoding/ormfield/duration.go:
```diff
@@ -1,11 +1,124 @@
+	seconds, nanos := getDurationSecondsAndNanos(value)
+	secondsInt := seconds.Int()
+	if secondsInt < DurationSecondsMin || secondsInt > DurationSecondsMax {
+		return fmt.Errorf("duration seconds is out of range %d, must be between %d and %d", secondsInt, DurationSecondsMin, DurationSecondsMax)
```
Might be nice to include a test case for expected error.

ryanchristo commented on orm/encoding/ormfield/timestamp.go:
```diff
@@ -21,38 +21,33 @@ import (
-	if secondsInt < timestampSecondsMin || secondsInt > timestampSecondsMax {
-		return fmt.Errorf("seconds is out of range %d, must be between %d and %d", secondsInt, timestampSecondsMin, timestampSecondsMax)
+	if secondsInt < TimestampSecondsMin || secondsInt > TimestampSecondsMax {
+		return fmt.Errorf("timestamp seconds is out of range %d, must be between %d and %d", secondsInt, TimestampSecondsMin, TimestampSecondsMax)
```
Might be nice to include a test case for expected error.

ryanchristo commented on orm/encoding/ormfield/timestamp.go:
```diff
@@ -63,65 +58,104 @@ func (t TimestampCodec) Encode(value protoreflect.Value, w io.Writer) error {
-	if nanosInt < 0 || nanosInt > timestampNanosMax {
-		return fmt.Errorf("nanos is out of range %d, must be between %d and %d", secondsInt, 0, timestampNanosMax)
+	if nanosInt < 0 || nanosInt > TimestampNanosMax {
+		return fmt.Errorf("timestamp nanos is out of range %d, must be between %d and %d", secondsInt, 0, TimestampNanosMax)
```
```suggestion
		return fmt.Errorf("timestamp nanos is out of range %d, must be between %d and %d", nanosInt, 0, TimestampNanosMax)
```

ryanchristo replied:
Might be nice to include a test case for expected error.

ryanchristo commented on orm/internal/testpb/test_schema.proto:
```diff
@@ -107,6 +107,18 @@ message ExampleTimestamp {
+    index: {id: 1 fields: "dur"}
+  };
+
+  uint64                    id   = 1;
+  string                    name = 2;
+  google.protobuf.Duration dur   = 3;
```
```suggestion
  uint64                   id   = 1;
  string                   name = 2;
  google.protobuf.Duration dur  = 3;
```

ryanchristo reviewed (commented):
Looking good. A handful of minor comments and suggestions.

tac0turtle commented:
@aaronc are you able to address the comments above?

aaronc commented:
@ryanchristo I addressed all your comments and added cases for all the out-of-range errors. I also generally refactored tests to be parallel and hopefully easier to follow

julienrbrt commented:
Somewhat unrelated, are we going to rename ORM go mod to the vanity url for v1? It's the only go mod that does not follow that.

aaronc commented:
> Somewhat unrelated, are we going to rename ORM go mod to the vanity url for v1? It's the only go mod that does not follow that.

Yes that makes sense
